### PR TITLE
Signal the worker progress event from the worker progress thread

### PR DESCRIPTION
**Pull request by pentschev**
Without signaling the worker progress event from the worker progress thread while shutting down the thread it may occur that the thread hangs as the event never gets awaken due to the lack of tasks to progress. This change ensures the worker progress event is signaled before joining the thread.

### CODE REVIEW

File: cpp/src/worker_progress_thread.cpp [hunk 1]
```diff
@@ -49,6 +53,7 @@ WorkerProgressThread::~WorkerProgressThread()
   }
 
   _stop = true;
+  _signalWorkerFunction();
   _thread.join();
 }
```
Comment (wence-) [inline cpp/src/worker_progress_thread.cpp:58]: Should the `_thread`'s destructor be called when deleting the worker progress thread?

File: cpp/src/worker.cpp [hunk 2]
```diff
@@ -254,19 +254,18 @@ void Worker::startProgressThread(const bool pollingMode)
...
                                                            _progressThreadStartCallbackArg,
                                                            _delayedSubmissionCollection);
 }
 
-void Worker::stopProgressThreadNoWarn()
-{
-  if (_progressThread && !_progressThread->pollingMode()) signal();
-  _progressThread = nullptr;
-}
+void Worker::stopProgressThreadNoWarn() { _progressThread = nullptr; }
```
Comment (wence-) [inline cpp/src/worker.cpp:268]: This doesn't need to signal now because `~WorkerProgressThread` arranges that for us?

**Review by wence- — APPROVED** [review]
I think this makes sense.

**Comment by wence-** [comment]
> I think this makes sense.

Although that said, it looks like the test suite [segfaulted](https://github.com/rapidsai/ucxx/actions/runs/4667908487/jobs/8264574976?pr=28#step:7:684)?

File: cpp/src/worker_progress_thread.cpp (hunk 1 above)
Comment (pentschev) [reply cpp/src/worker_progress_thread.cpp:58]: Not sure I understand the question. The `_thread` destructor will run when the `~WorkerProgressThread` goes out-of-scope, did you expect something different?

File: cpp/src/worker.cpp (hunk 2 above)
Comment (pentschev) [reply cpp/src/worker.cpp:268]: Precisely, it must be signaled before we join the thread, which can't be done from here. In any case, I think it makes more sense that `~WorkerProgressThread` performs this operation too, making it less error prone.

**Comment by pentschev** [comment]
> Although that said, it looks like the test suite [segfaulted](https://github.com/rapidsai/ucxx/actions/runs/4667908487/jobs/8264574976?pr=28#step:7:684)?

It didn't segfault, it timed out. I can reproduce that locally and it's hitting https://github.com/rapidsai/ucxx/issues/15 , although in CI it seems to hang indefinitely. That issue is not related to this PR specifically and is happening since we added CI. I'm still investigating it but going to go ahead and merge this when CI passes, as it fixed timeouts when running C++ examples.

Thanks @wence- !

**Pull request merged**